cmd/kube-apiserver/app/testing/testserver.go

```diff
@@ -114,14 +116,41 @@ func StartTestServer(t Logger, instanceOptions *TestServerInstanceOptions, custo
 	for _, f := range s.Flags().FlagSets {
 		fs.AddFlagSet(f)
 	}
-
 	s.InsecureServing.BindPort = 0
 
 	s.SecureServing.Listener, s.SecureServing.BindPort, err = createLocalhostListenerOnFreePort()
 	if err != nil {
 		return result, fmt.Errorf("failed to create listener: %v", err)
 	}
 	s.SecureServing.ServerCert.CertDirectory = result.TmpDir
+
+	// create optional certificates for aggregation and client-cert auth
+	proxySigningKey, err := testutil.NewPrivateKey()
+	if err != nil {
+		return result, err
+	}
+	proxySigningCert, err := cert.NewSelfSignedCACert(cert.Config{CommonName: "front-proxy-ca"}, proxySigningKey)
+	if err != nil {
+		return result, err
+	}
+	proxyCACertFile := path.Join(s.SecureServing.ServerCert.CertDirectory, "proxy-ca.crt")
+	if err := ioutil.WriteFile(proxyCACertFile, testutil.EncodeCertPEM(proxySigningCert), 0644); err != nil {
+		return result, err
+	}
+	s.Authentication.RequestHeader.ClientCAFile = proxyCACertFile
+	clientSigningKey, err := testutil.NewPrivateKey()
+	if err != nil {
+		return result, err
+	}
+	clientSigningCert, err := cert.NewSelfSignedCACert(cert.Config{CommonName: "client-ca"}, clientSigningKey)
+	if err != nil {
+		return result, err
+	}
+	clientCACertFile := path.Join(s.SecureServing.ServerCert.CertDirectory, "client-ca.crt")
+	if err := ioutil.WriteFile(clientCACertFile, testutil.EncodeCertPEM(clientSigningCert), 0644); err != nil {
+		return result, err
+	}
+	s.Authentication.ClientCert.ClientCA = clientCACertFile
```

Comment:
This plumbing is against the design of the testserver. Use flags to configure. We only do direct options manipulation where it is absolute necessary.

Reply:
> This plumbing is against the design of the testserver. Use flags to configure. We only do direct options manipulation where it is absolute necessary.

External callers set flags, but all the internal options are set via options.  See a couple lines above where directories with flags are explicitly set and a few lines down where runtimeconfigs are directly set.  The only flags are set by external callers, but these options just complete "normal" wiring authentication that callers don't need to manipulate.